File: csharp/Link.Foundation.Links.Notation.Tests/StreamParserTests.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Xunit;

namespace Link.Foundation.Links.Notation.Tests
{
    public static class StreamParserTests
    {
        private const string Document =
            "first loves data\n" +
            "# streamed comment\n" +
            "profile:\n" +
            "  name Ada\n" +
            "  note \"line one\nline two\"\n" +
            "(nested:\n  child value)\n" +
            "last sees first";

        private static string[] Render(IEnumerable<Link<string>> links) =>
            links.Select(link => link.ToString()).ToArray();

        [Fact]
        public static void MatchesCanonicalParserOneSymbolAtATime()
        {
            var stream = new StreamParser();
            foreach (var character in Document)
            {
                stream.Write(character.ToString());
            }

            Assert.Equal(Render(new Parser().Parse(Document)), Render(stream.Finish()));
        }

        [Fact]
        public static void EmitsOnlyCompleteRecords()
        {
            var seen = new List<Link<string>>();
            var stream = new StreamParser();
            stream.LinkParsed += seen.Add;

            Assert.Empty(stream.Write("profile:\n  name Ada\n  note \"first\nsecond\"\n"));
            var emitted = stream.Write("n");

            Assert.Single(emitted);
            Assert.Equal(emitted, seen);
        }

        [Fact]
        public static void SupportsLineChunksFinalRecordAndPosition()
        {
            var stream = new StreamParser();
            foreach (var line in Document.SplitAfter('\n'))
            {
                stream.Write(line);
            }

            Assert.Equal(Render(new Parser().Parse(Document)), Render(stream.Finish()));
            Assert.Equal(Document.Length, stream.Position.Offset);
            Assert.Equal(0, stream.Position.Buffered);
        }

        [Fact]
        public static void SupportsDrainResetAndBoundedMemory()
        {
            var seen = new List<Link<string>>();
            var stream = new StreamParser { Collect = false, MaxBufferSize = 8 };
            stream.LinkParsed += seen.Add;
            for (var index = 0; index < 100; index++) stream.Write("a\n");

            Assert.Empty(stream.Drain());
            Assert.Single(stream.Finish());
            Assert.Equal(100, seen.Count);

            stream.Reset();
            Assert.Throws<ArgumentException>(() => { stream.Write("123456789"); });
        }

        [Fact]
        public static void ProvidesLazyAdapters()
        {
            var actual = StreamParser.ParseChunks(new[] { "one link\n", "two link" });

            Assert.Equal(new[] { "(one link)", "(two link)" }, Render(actual));
        }

        [Fact]
        public static void RejectsWritesAfterFinish()
        {
            var stream = new StreamParser();
            stream.Finish("one");

            Assert.Throws<InvalidOperationException>(() => stream.Write("two"));
        }

        [Fact]
        public static async Task ProvidesAnAsyncEnumerableAdapter()
        {
            static async IAsyncEnumerable<string> Chunks()
            {
                yield return "one link\n";
                await Task.Yield();
                yield return "two link";
            }

            var actual = new List<Link<string>>();
            await foreach (var link in StreamParser.ParseChunksAsync(
                Chunks(),
                cancellationToken: TestContext.Current.CancellationToken)) actual.Add(link);

            Assert.Equal(new[] { "(one link)", "(two link)" }, Render(actual));
        }
    }
}

internal static class StreamingTestStringExtensions
{
    public static IEnumerable<string> SplitAfter(this string value, char delimiter)
    {
        var start = 0;
        for (var index = 0; index < value.Length; index++)
        {
            if (value[index] != delimiter) continue;
            yield return value.Substring(start, index - start + 1);
            start = index + 1;
        }
        if (start < value.Length) yield return value.Substring(start);
    }
}

File: rust/links-notation/tests/stream_parser_tests.rs
use links_notation::{format_links, parse_lino_to_links, StreamParser};
use std::sync::{Arc, Mutex};

const DOCUMENT: &str = "first loves data\n\
profile:\n\
  name Ada\n\
  note \"line one\nline two\"\n\
(nested:\n  child value)\n\
last sees first";

#[test]
fn matches_the_canonical_parser_one_character_at_a_time() {
    let mut stream = StreamParser::new();
    for character in DOCUMENT.chars() {
        stream.write(&character.to_string()).unwrap();
    }

    assert_eq!(
        format_links(&stream.finish().unwrap()),
        format_links(&parse_lino_to_links(DOCUMENT).unwrap())
    );
}

#[test]
fn emits_only_complete_records_to_the_callback() {
    let seen = Arc::new(Mutex::new(Vec::new()));
    let callback_seen = Arc::clone(&seen);
    let mut stream = StreamParser::new();
    stream.on_link(move |link| callback_seen.lock().unwrap().push(link.to_string()));

    assert!(stream
        .write("profile:\n  name Ada\n  note \"first\nsecond\"\n")
        .unwrap()
        .is_empty());
    let emitted = stream.write("n").unwrap();

    assert_eq!(emitted.len(), 1);
    assert_eq!(*seen.lock().unwrap(), vec![emitted[0].to_string()]);
}

#[test]
fn supports_line_chunks_final_record_and_position() {
    let mut stream = StreamParser::new();
    for line in DOCUMENT.split_inclusive('\n') {
        stream.write(line).unwrap();
    }

    assert_eq!(
        format_links(&stream.finish().unwrap()),
        format_links(&parse_lino_to_links(DOCUMENT).unwrap())
    );
    assert_eq!(stream.position().offset, DOCUMENT.len());
    assert_eq!(stream.position().buffered, 0);
}

#[test]
fn supports_drain_reset_and_bounded_memory() {
    let seen = Arc::new(Mutex::new(0));
    let callback_seen = Arc::clone(&seen);
    let mut stream = StreamParser::new();
    stream.set_collect(false);
    stream.set_max_buffer_size(8).unwrap();
    stream.on_link(move |_| *callback_seen.lock().unwrap() += 1);
    for _ in 0..100 {
        stream.write("a\n").unwrap();
    }

    assert!(stream.drain().is_empty());
    assert_eq!(stream.finish().unwrap().len(), 1);
    assert_eq!(*seen.lock().unwrap(), 100);

    stream.reset();
    assert!(stream.write("123456789").is_err());
}

#[test]
fn provides_an_iterator_adapter() {
    let actual = StreamParser::parse_chunks(["one link\n", "two link"])
        .map(|link| link.unwrap().to_string())
        .collect::<Vec<_>>();

    assert_eq!(actual, vec!["(one link)", "(two link)"]);
}
